android/libraries/rib-badoo/src/main/java/com/badoo/common/rib/routing/backstack/BackStackManager.kt

```diff
@@ -0,0 +1,187 @@
+
+internal class BackStackManager<C : Parcelable>(
+    resolver: (C) -> RoutingAction<*>,
+    ribConnector: RibConnector,
+    initialConfiguration: C,
+    timeCapsule: TimeCapsule<State<C>>,
+    tag: String
+): BaseFeature<Wish<C>, Action<C>, Effect<C>, State<C>, Nothing>(
+    initialState = timeCapsule[tag] ?: State(),
+    wishToAction = { Execute(it) },
+    bootstrapper = BooststrapperImpl(timeCapsule[tag] ?: State(), initialConfiguration),
+    actor = ActorImpl<C>(
+        connector = BackStackRibConnector(
+            resolver,
+            ribConnector
+        )
+    ),
+    reducer = ReducerImpl<C>()
+) {
+    init {
+        timeCapsule.register(tag) { state }
+    }
+
+    @Parcelize
+    data class State<C : Parcelable>(
+        val backStack: List<BackStackElement<C>> = emptyList()
+    ) : Parcelable {
+        val current: C
+            get() = backStack.last().configuration
+
+        val canPop: Boolean
+            get() = backStack.size > 1
+    }
+
+    class BooststrapperImpl<C : Parcelable>(
+        private val state: State<C>,
+        private val initialConfiguration: C
+    ) : Bootstrapper<Action<C>> {
+        override fun invoke(): Observable<Action<C>> = when {
+            state.backStack.isEmpty() -> just(Execute(NewRoot(initialConfiguration)))
+            else -> just(ReattachRibsOfLastEntry())
+        }
+    }
+
+    sealed class Wish<C : Parcelable> {
+        data class Replace<C : Parcelable>(val configuration: C) : Wish<C>()
+        data class Push<C : Parcelable>(val configuration: C) : Wish<C>()
+        data class NewRoot<C : Parcelable>(val configuration: C) : Wish<C>()
+        class Pop<C : Parcelable> : Wish<C>()
+        class ShrinkToBundles<C : Parcelable> : Wish<C>()
+        class TearDown<C : Parcelable> : Wish<C>()
+    }
+
+    sealed class Action<C : Parcelable> {
+        data class Execute<C : Parcelable>(val wish: Wish<C>) : Action<C>()
+        class ReattachRibsOfLastEntry<C : Parcelable> : Action<C>()
+    }
+
+    sealed class Effect<C : Parcelable> {
+        data class Replace<C : Parcelable>(val newEntry: BackStackElement<C>) : Effect<C>()
+        data class Push<C : Parcelable>(val updatedOldEntry: BackStackElement<C>, val newEntry: BackStackElement<C>) : Effect<C>()
+        data class NewRoot<C : Parcelable>(val newEntry: BackStackElement<C>) : Effect<C>()
+        data class Pop<C : Parcelable>(val updatedOldEntry: BackStackElement<C>) : Effect<C>()
+        data class UpdateBackStack<C : Parcelable>(val updatedBackStack: List<BackStackElement<C>>) : Effect<C>()
+    }
+
+    class ActorImpl<C : Parcelable>(
+        private val connector: BackStackRibConnector<C>
+    ) : Actor<State<C>, Action<C>, Effect<C>> {
+
+        override fun invoke(state: State<C>, action: Action<C>): Observable<out Effect<C>> =
+            when (action) {
+                is ReattachRibsOfLastEntry -> Observable.defer {
+                    return@defer just(
```

Comment:
Not sure you need `defer` here.
You can use Completable.fromAction { } and then use `toObservable`on it